src/connect.tsx

```diff
@@ -0,0 +1,163 @@
+    // soon (but still async) by re-rendering after a router change would otherwise receive
+    // the new (and possibly incorrect from their POV) state
+    // TODO(gio): elaborate better / try to remove
+    //
+    // Boolean
+    //
+    filterValid = true,
+
+    // optionally kill some props using lodash omit
+    // useful to stay strict in @props declaration, a typical default could be:
+    // ['router', 'query', 'params']
+    //
+    // Array<String>
+    //
+    killProps = []
+  }: ConnectConfig = {}) {
+
+    if (process.env.NODE_ENV !== 'production') {
+      const isKeyList = t.list(t.String).is(decl);
+      if (!isKeyList) {
+        throw new Error('connect expects a list of state keys');
+      }
+
+      decl.forEach(k => {
+        if (!stateType.meta.props.hasOwnProperty(k)) {
+          throw new Error(`${k} is not defined in state`);
+        }
+      });
+    }
+
+    const decorator: ConnectDecorator<S, Decl> = <P extends {}>(Component: DecoratedComponent<S, P>) => {
+      const displayName = `connect(${Component.displayName || Component.name || 'Component'})`;
+
+      const getNewState = defaultGetNewState(decl);
+
+      const shouldUpdateState = filterValid ? (v: S) => {
+        const invalid = decl.reduce((acc, k) => {
+          return !(stateType.meta.props[k] as t.Type<any>).is(v[k]) ? [...acc, k] : acc; // TODO(typo)
+        }, []);
+
+        if (invalid.length > 0 && process.env.NODE_ENV === 'development') {
+          warn(`Skipping update for ${displayName}. Invalid keys: ${invalid.join(', ')}`);
+        }
+        return invalid.length === 0;
+      } : () => true;
+
+      type ContextT<S extends ST> = { // TODO(typo): move elsewhere and reuse
+        state: BehaviorSubject<S>
+        transition: TransitionFunction<S>
+      }
+      type StateT<S extends ST, Decl extends keyof S> = { [k in Decl]: S[k] } | {}
+      return class ConnectWrapper extends React.Component<ConnectedComponentProps<P, Decl>, StateT<S, Decl>> {
+        static contextTypes = ConnectContextTypes
+
+        static displayName = displayName;
+
+        constructor(props: ConnectedComponentProps<P, Decl>, context: ContextT<S>) {
+          super(props, context);
+          const value = context.state.value;
+          if (shouldUpdateState(value)) {
+            this.state = getNewState(value);
+          } else {
+            this.state = {};
+          }
+        }
+
+        context: ContextT<S>
+        _subscription?: Subscription
+
+        componentDidMount() {
+          this._subscription = this.context.state.filter(shouldUpdateState).map(getNewState).subscribe(this.setState.bind(this));
+        }
+
+        componentWillUnmount() {
+          if (this._subscription) {
+            this._subscription.unsubscribe();
+          }
+        }
+
+        shouldComponentUpdate(nextProps: ConnectedComponentProps<P, Decl>, nextState: StateT<S, Decl>) {
```

Comment:
not super easy to do because it is based on a config parameter (`pure`).. I think we'd need two different declarations. I'll leave it as is for now